### ISSUE TITLE
Templated policies for Kubernetes Auth

### ISSUE THREAD

**Issue by andre-taveira** [hashicorp/vault#7497]
**Is your feature request related to a problem? Please describe.**

I have a Kubernetes cluster that I want to share with an Engineering team. I want them to be able to write apps that can read and/or write data to vault using Kubernetes Service account tokens that can be used to authenticate to Vault (Kubernetes Auth enabled). The caveat here is that I'd like to ensure that each app can only change or read their own data in vault in a path that represents the namespace where they live which is a metadata already associated with the token Vault generates for a k8s auth.

**Describe the solution you'd like**

It would be nice if I could write a templated policy like:

/# Create a policy to give read-write privileges only in a path that represents the namespace from the service-account used to access vault. Something like:
path "my-cluster-name/{{auth.kubernetes.token_meta_service_account_namespace}}/*" {
  capabilities = ["create", "update", "read", "delete"]
}

/#Write the policy
vault policy write -tls-skip-verify "k8s-namespace-read-write" k8s-namespace-read-write.hcl

/#bind the policy to service_account in K8s with the vault-read-write name
vault write -tls-skip-verify auth/my-cluster-name/role/k8s-app-rw-vault \
    bound_service_account_names=vault-read-write \
    bound_service_account_namespaces="*" \
    policies="k8s-namespace-read-write" \
    ttl=1h

/# What vault returns when I authenticate with my Service Account token
vault write -tls-skip-verify auth/my-cluster-name/login role=k8s-app-rw-vault -path=/andre-cluster jwt=SERVICE_ACCOUNT_TOKEN

Key                                       Value
---                                       -----
token                                     ade1ecb1-fa02-84e2-e863-18270be3a367
token_accessor                            28acd6b6-efab-8df1-4b35-3c2b99b490a2
token_duration                            1h
token_renewable                           true
token_policies                            []
identity_policies                         []
policies                                  ["default" "k8s-namespace-read-write"]
token_meta_service_account_name           vault-read-write
token_meta_service_account_namespace      read-write-vault-test
token_meta_service_account_secret_name    vault-read-write-token-n4l76
token_meta_service_account_uid            3faeb6f3-db0f-11e9-9d2d-0a13ca37e426
token_meta_role                           k8s-app-rw-vault

**Describe alternatives you've considered**

As far as I understand the templated policies work only for entities usage. My other alternative would be to use a vault-service-broker (https://github.com/hashicorp/vault-service-broker) and deploy my kubernetes resources with a request to get a token with access to a specific path. The service broker would them create that path , add policies and return me the token to use. Adds lots of complexity.

**Explain any additional use-cases**

This would help us achieve a very simple multi-tenant usage of vault for a shared kubernetes cluster

**Comment by briankassouf** [comment]
I believe this is already possible using the identity templates. Entities are automatically created on login so using the entity alias associated with this mount should work. Something like:
```
path "my-cluster-name/{{identity.entity.aliases.<<mount accessor>>.metadata.service_account_namespace}}/*" {
    capabilities = ["create", "update", "read", "delete"]
}
```
You'll have to replace `<<mount accessor>>` with the actual mount accessor from the `vault secrets list -detailed` output.

**Comment by jefferai** [comment]
Correct, it is not available in that version. There have also been numerous important bug and security fixes -- I strongly recommend upgrading.

**Comment by nouseforaname** [comment]
Hi,

we updated the vault instances. Unfortunately we still cannot make it work. The above config still applies

**Comment by briankassouf** [comment]
@nouseforaname I think it should be:
 ```
test-cluster/{{identity.entity.aliases.auth_kubernetes_606fd930.metadata.service_account_namespace}}/*
```

Notice the double `}}/*` at the end.

**Comment by nouseforaname** [comment]
That was a copy paste error.

**Comment by briankassouf** [comment]
Gotcha, what error are you seeing now? Can you layout the steps you're taking?

**Comment by nouseforaname** [comment]
I do not have access to the env anymore. Vault is not in use too if I remember correctly.. @andre-taveira should know

**Comment by loosi** [comment]
It works with version 1.3.1 
```
path "foooobardata/data/$CLUSTERNAME/{{identity.entity.aliases.auth_kubernetes_b4599e21.metadata.service_account_namespace}}/*"
```
auth_kubernetes_b4599e21 = is our mount_accesor for the specific auth mount (one of the kube clusters)

**Comment by andre-taveira** [comment]
Hey guys ! thanks for the update ! We just resumed these tests. I'll let you know soon our test results :)

**Comment by elthariel** [comment]
TL;DR: The `{{identity.entity.aliases.auth_kubernetes_b4599e21.metadata.service_account_namespace}}` templating method works.
You can get the kube auth accessor using `vault auth list`

I was trying to achieve the very same thing when I was lucky enough to find this issue, because I wouldn't have guessed it, and would have to dive into the code to find out.

My problem applying this method was to find the name of the actual mount accessor, as it wasn't listed in the `vault secrets list -detailed` output. I've been using the raw audit log to find the token accessor, then the identity api to recover the mount accessor... before I realized it was there, under my nose in the `vault auth list` command 😭 .

That would be really awesome to drop a few lines in the documentation about that. I'll see if I can send a PR

**Comment by briankassouf** [comment]
Just merged @elthariel's docs PR for this and since it seems to be working for everyone i'll close this issue out.
